File: src/python/OrderBook.py
from collections import deque
import datetime as dt
import json
import math
import random
from Order import Order, OrderAck, OrderStatus, Fill
from DataEvent import DataEvent, AsyncDataEvent
import asyncio
from itertools import groupby
from operator import attrgetter
import traceback
import sys

class OrderBook:
    def __init__(self, symbol):
        self.symbol = symbol
        self.websocket = []
        self.loop = asyncio.get_event_loop()
        # self.send_book_event = AsyncDataEvent()
        self.send_book_event = AsyncDataEvent()
        self.send_trade_event = AsyncDataEvent()
        self.new_event = AsyncDataEvent()
        self.modify_event = AsyncDataEvent()
        self.cancel_event = AsyncDataEvent()
        self.bids = {}
        self.asks = {}
        self.ordersById = {}

    async def start(self):
        waitNewTask = asyncio.create_task(self.wait_new(self.new_event))
        waitModifyTask = asyncio.create_task(self.wait_modify(self.modify_event))
        waitCancelTask = asyncio.create_task(self.wait_cancel(self.cancel_event))
        waitSendBookTask = asyncio.create_task(self.wait_send_book(self.send_book_event))
        waitSendTradeTask = asyncio.create_task(self.wait_send_trade(self.send_trade_event))
        # # await asyncio.gather(waitNewTask, waitModifyTask, waitCancelTask, waitSendBookTask)

    def toJson(self, msgType, fill=None):
        bids = sum([list(o) for o in [self.bids[lvl] for lvl in sorted(list(self.bids.keys()))[::-1]]], [])
        asks = sum([list(o) for o in [self.asks[lvl] for lvl in sorted(list(self.asks.keys()))]], [])

        if msgType == 'W':
            bidPxs = sorted(list(set([o.px for o in set(bids)])))[::-1]
            askPxs = sorted(list(set([o.px for o in set(asks)])))
            # if len(askPxs) > 1:
            #     print(self.asks)
            bidQtys = [sum(o.qty for o in groupedOs) for px, groupedOs in groupby(bids, attrgetter('px'))]
            askQtys = [sum(o.qty for o in groupedOs) for px, groupedOs in groupby(asks, attrgetter('px'))]
            lvlCount = len(bidPxs) + len(askPxs)
            fix42 = {35:'W',52:str(dt.datetime.now()),55:self.symbol,268:[]}
            for i in range(len(bidPxs)):
                fix42[268].append({269:0,270:bidPxs[i],271:bidQtys[i]})
            for i in range(len(askPxs)):
                fix42[268].append({269:1,270:askPxs[i],271:askQtys[i]})
        # elif msgType == 'X':
        #     fix42 = { 35:'X'
        #              ,269:2
        #              ,52:str(fill.date)
        #              ,55:self.symbol
        #              ,56:fill.clientId
        #              ,54:(1 if fill.side == "Buy" else 2)
        #              ,38:fill.qty
        #              ,44:fill.px}

        return json.dumps(fix42)

    async def wait_new(self, e):
        print("Awaiting new orders")
        while True:
            await e.waitRun(self.newOrder)
            e.clear()

    async def newOrder(self, o):
        self.ordersById[o.id] = o
        side, other_side = (self.bids, self.asks) if o.side == 'Buy' else (self.asks, self.bids)
        if o.px not in side:
            side[o.px] = deque()
        side[o.px].append(o)
        if o.ackEvent:
            print(f"{dt.datetime.now().timestamp()}:: OB:Setting New Order AckEvent: {o.ackEvent}")
            try:
                o.ackEvent.set(OrderAck(o,OrderStatus.NEW))
            except Exception as ex:
                print(ex)
                traceback.print_exc(file=sys.stdout)
            finally:
                o.ackEvent.clear()
        self.match(o)
        self.send_book_event.set()

    async def wait_modify(self, e):
        print("Awaiting modifies")
        while True:
            await e.waitRun(self.modifyOrder)
            e.clear()

    async def modifyOrder(self, data):
        try:
            (orderId, px, qty) = data
            if orderId not in self.ordersById:
                return
            o = self.ordersById[orderId]
            if qty is None: qty = o.qty
            side, other_side = (self.bids, self.asks) if o.side == 'Buy' else (self.asks, self.bids)
            backOfTheLine = px != o.px or qty > o.qty
            if backOfTheLine and o.px in side:
                side[o.px].remove(o)
                if len(side[o.px]) == 0:
                    del side[o.px]
            o.px = px
            o.qty = qty
            if backOfTheLine and o.id in self.ordersById:
                if o.px not in side:
                    side[o.px] = deque()
                side[o.px].append(o)
        except Exception as ex:
            print(ex)
            traceback.print_exc(file=sys.stdout)
        if o.ackEvent:
            print(f"{dt.datetime.now().timestamp()}:: OB:Setting Modify AckEvent: {o}")
            o.ackEvent.set(OrderAck(o,OrderStatus.MODIFIED))
        self.match(o)
        self.send_book_event.set()

    async def wait_cancel(self, e):
        print("Awaiting cancels")
        while True:
            await e.waitRun(self.cancelOrder)
            e.clear()
        
    async def cancelOrder(self, orderId):
        print(f'Cancel order {orderId}')
        o = self.ordersById[orderId]
        del self.ordersById[o.id]
        side, other_side = (self.bids, self.asks) if o.side == 'Buy' else (self.asks, self.bids)
        side[o.px].remove(o)
        if len(side[o.px]) == 0:
            del side[o.px]
        if o.ackEvent:
            o.ackEvent.set(OrderAck(o,OrderStatus.CANCELLED))
        self.send_book_event.set()
            
    def fill(self, o, sz):
        print(f'OB: Fill')
        exec_time = dt.datetime.now()
        side, other_side = (self.bids, self.asks) if o.side == 'Buy' else (self.asks, self.bids)
        o.qty -= sz
        pre = 'PARTIAL'
        if o.qty == 0:
            del self.ordersById[o.id]
            pre = 'FULL'
            # print(f'Removing {o} from {o.side} orders')
            side[o.px].remove(o)
            if len(side[o.px]) == 0:
                # print(f'Removing {o.px} from side: {o.side}')
                del side[o.px]
            if o.ackEvent:
                print('OB: Full fill')
                o.ackEvent.set(OrderAck(o,OrderStatus.FULLY_FILLED))
        else:
            if o.ackEvent:
                o.ackEvent.set(OrderAck(o,OrderStatus.PARTIALLY_FILLED))
        self.send_trade_event.set(Fill(exec_time, o.id, o.id, o.clientId, o.side, sz, o.px, o.qty == 0))
        print(f'{pre} FILL: {o.clientId} {o.side}s {sz} @ {o.px} {o.qty} Remaining')
        self.send_book_event.set()
        

    def match(self, o):
        side, other_side = (self.bids, self.asks) if o.side == 'Buy' else (self.asks, self.bids)
        # orders = side[o.px]
        matching_orders = [self.asks[k] for k in sorted(list(self.asks.keys())) if k <= o.px] if o.side == 'Buy' else [self.bids[k] for k in sorted(list(self.bids.keys())) if k >= o.px]
        matching_orders = [order for orders in matching_orders for order in orders]
        found_matches = len(matching_orders) > 0
        # if found_matches:
        #     print(f'o:{o} matching_orders:{matching_orders}')
        i = 0
        while o.qty > 0 and i < len(matching_orders):
            mo = matching_orders[i]
            print(f'{o} matched with {mo}')
            print(self)
            sz = min(o.qty, mo.qty)
            # full_fill_o = o.qty - sz == 0
            self.fill(o,sz)
            # print(f'After FILL, o.qty = {o.qty}')
            full_fill_mo = mo.qty - sz == 0
            self.fill(mo, sz)
            if full_fill_mo:
                i += 1
            print(self)

    async def wait_send_book(self, e):
        print(f'Waiting to send book')
        while True:
            await e.waitRun(self.send_book)
            e.clear()

    async def send_book(self, data=None):
        msg = self.toJson('W')
        if self.websocket:
            for ws in self.websocket:
                if ws.open:
                    print(f"{dt.datetime.now().timestamp()}:: OB:Sending Book")
                    await ws.send(msg)

    async def wait_send_trade(self, e):
        print(f'Waiting to send trades')
        while True:
            await e.waitRun(self.send_trade)
            e.clear()

    async def send_trade(self, fill=None):
        pass;
        # msg = self.toJson('X', fill=fill)
        # if self.websocket:
        #     for ws in self.websocket:
        #         if ws.open:
        #             await ws.send(msg)

    def __str__(self):
        # return "OrderBook"
        return f'bids: {self.bids}\nasks: {self.asks}'

    def __repr__(self):
        return str(self)

File: src/components/OrderBook.js
import { AsyncDataEvent } from "./DataEvent";
import { OrderAck, OrderStatus, Fill } from './Order';

class OrderBook {
    constructor(symbol, ms_send) {
        this.symbol = symbol;
        this.ms_send = ms_send
        this.loop = null; // JavaScript doesn't have a direct equivalent to asyncio.get_event_loop()
        this.send_book_event = new AsyncDataEvent();
        this.send_trade_event = new AsyncDataEvent();
        this.new_event = new AsyncDataEvent();
        this.modify_event = new AsyncDataEvent();
        this.cancel_event = new AsyncDataEvent();
        this.bids = {};
        this.asks = {};
        this.ordersById = {};
    }

    async start() {
        console.log("OB: Starting OrderBook");
        this.waitNewPromise = this.wait_new(this.new_event);
        this.waitModifyPromise = this.wait_modify(this.modify_event);
        this.waitCancelPromise = this.wait_cancel(this.cancel_event);
        this.waitSendBookPromise = this.wait_send_book(this.send_book_event);
        this.waitSendTradePromise = this.wait_send_trade(this.send_trade_event);
    }

    toJson(msgType, fill = null) {
        const bids = [].concat(...Object.keys(this.bids).sort((a, b) => b - a).map(lvl => this.bids[lvl]));
        const asks = [].concat(...Object.keys(this.asks).sort().map(lvl => this.asks[lvl]));

        let fix42 = { 35: 'W', 52: new Date().toISOString(), 55: this.symbol, 268: [] };
        if (msgType === 'W') {
            const bidPxs = [...new Set(bids.map(o => o.px))].sort((a, b) => b - a);
            const askPxs = [...new Set(asks.map(o => o.px))].sort();
            const bidQtys = bidPxs.map(px => bids.filter(o => o.px === px).reduce((sum, o) => sum + o.qty, 0));
            const askQtys = askPxs.map(px => asks.filter(o => o.px === px).reduce((sum, o) => sum + o.qty, 0));
            bidPxs.forEach((px, i) => {
                fix42[268].push({ 269: 0, 270: px, 271: bidQtys[i] });
            });
            askPxs.forEach((px, i) => {
                fix42[268].push({ 269: 1, 270: px, 271: askQtys[i] });
            });
        }

        return fix42;
    }

    async wait_new(e) {
        console.log("Awaiting new orders");
        while (true) {
            await e.waitRun(this.newOrder.bind(this));
            // e.clear();
        }
    }

    async newOrder(o) {
        console.log("OB: newOrder: ", o);
        o.status = OrderStatus.NEW;
        this.ordersById[o.id] = o;
        const [side, other_side] = o.side === 'Buy' ? [this.bids, this.asks] : [this.asks, this.bids];
        if (!side[o.px]) {
            side[o.px] = [];
        }
        side[o.px].push(o);
        if (o.ackEvent) {
            console.log(`${new Date().toISOString()}:: OB:Setting New Order AckEvent: ${o.ackEvent}`);
            try {
                o.ackEvent.set(new OrderAck(o, OrderStatus.NEW));
            } catch (ex) {
                console.error(ex);
                console.trace();
            } 
            // finally {
            //     o.ackEvent.clear();
            // }
        }
        this.match(o);
        this.send_book_event.set();
    }

    async wait_modify(e) {
        console.log("Awaiting modifies");
        while (true) {
            await e.waitRun(this.modifyOrder.bind(this));
            // e.clear();
        }
    }

    async modifyOrder(data) {
        try {
            // console.log("OB: modifyOrder: ", data);
            let [orderId, px, qty] = data;
            if (!this.ordersById[orderId]) return;
            const o = this.ordersById[orderId];
            if (qty === null) qty = o.qty;
            const [side, other_side] = o.side === 'Buy' ? [this.bids, this.asks] : [this.asks, this.bids];
            const backOfTheLine = px !== o.px || qty > o.qty;
            if (backOfTheLine && o.px in side) {
                side[o.px] = side[o.px].filter(order => order !== o);
                if (side[o.px].length === 0) {
                    delete side[o.px];
                }
            }
            o.px = px;
            o.qty = qty;
            o.status = OrderStatus.MODIFIED;
            if (backOfTheLine && o.id in this.ordersById) {
                if (!(o.px in side)) {
                    side[o.px] = [];
                }
                side[o.px].push(o);
            }
            if (o.ackEvent) {
                o.ackEvent.set(new OrderAck(o, OrderStatus.MODIFIED));
            }
            this.match(o);
            this.send_book_event.set();
        } catch (ex) {
            console.error(ex);
            console.trace();
        }
    }

    async wait_cancel(e) {
        console.log("Awaiting cancels");
        while (true) {
            await e.waitRun(this.cancelOrder.bind(this));
            // e.clear();
        }
    }

    async cancelOrder(orderId) {
        console.log(`Cancel order ${orderId}`);
        const o = this.ordersById[orderId];
        o.status = OrderStatus.CANCELLED;
        delete this.ordersById[o.id];
        const [side, other_side] = o.side === 'Buy' ? [this.bids, this.asks] : [this.asks, this.bids];
        side[o.px] = side[o.px].filter(order => order !== o);
        if (side[o.px].length === 0) {
            delete side[o.px];
        }
        if (o.ackEvent) {
            o.ackEvent.set(new OrderAck(o, OrderStatus.CANCELLED));
        }
        this.send_book_event.set();
    }

    fill(o, sz) {
        console.log('OB: Fill');
        const exec_time = new Date();
        const [side, other_side] = o.side === 'Buy' ? [this.bids, this.asks] : [this.asks, this.bids];
        o.qty -= sz;
        o.filled_qty += sz;
        let pre = 'PARTIAL';
        if (o.qty === 0) {
            o.status = OrderStatus.FULLY_FILLED;
            delete this.ordersById[o.id];
            pre = 'FULL';
            side[o.px] = side[o.px].filter(order => order !== o);
            if (side[o.px].length === 0) {
                delete side[o.px];
            }
            if (o.ackEvent) {
                console.log('OB: Full fill', o);
                o.ackEvent.set(new OrderAck(o, OrderStatus.FULLY_FILLED));
            }
        } else {
            o.status = OrderStatus.PARTIALLY_FILLED;
            if (o.ackEvent) {
                o.ackEvent.set(new OrderAck(o, OrderStatus.PARTIALLY_FILLED));
            }
        }
        this.send_trade_event.set(new Fill(exec_time, o.id, o.id, o.clientId, o.side, sz, o.px, o.qty === 0));
        console.log(`${pre} FILL: ${o.clientId} ${o.side}s ${sz} @ ${o.px} ${o.qty} Remaining`);
        this.send_book_event.set();
    }

    match(o) {
        const [side, other_side] = o.side === 'Buy' ? [this.bids, this.asks] : [this.asks, this.bids];
        const matching_orders = o.side === 'Buy' ?
            Object.keys(this.asks).filter(k => k <= o.px).flatMap(k => this.asks[k]) :
            Object.keys(this.bids).filter(k => k >= o.px).flatMap(k => this.bids[k]);
        let found_matches = matching_orders.length > 0;
        let i = 0;
        while (o.qty > 0 && i < matching_orders.length) {
            const mo = matching_orders[i];
            console.log(`${o} matched with ${mo}`);
            const sz = Math.min(o.qty, mo.qty);
            this.fill(o, sz);
            const full_fill_mo = mo.qty - sz === 0;
            this.fill(mo, sz);
            if (full_fill_mo) {
                i++;
            }
        }
    }

    async wait_send_book(e) {
        console.log('OB: Waiting to send book');
        while (true) {
            try {
                await e.waitRun(this.send_book.bind(this));
            } catch (ex) {
                console.error(ex);
                console.trace();
            }
            // e.clear();
        }
    }

    async send_book(data = null) {
        const msg = this.toJson('W');
        this.ms_send(msg);
    }

    async wait_send_trade(e) {
        console.log('Waiting to send trades');
        while (true) {
            await e.waitRun(this.send_trade.bind(this));
            // e.clear();
        }
    }

    async send_trade(fill = null) {
        // Implementation here is commented out in Python version
    }

    toString() {
        return `bids: ${JSON.stringify(this.bids)}\nasks: ${JSON.stringify(this.asks)}`;
    }
}

export default OrderBook;